### PR TITLE
Print pdf inside and outside kDrive

### PR DESCRIPTION
**Pull request by tevincent**
https://github.com/Infomaniak/android-kDrive/assets/149579879/73a08f10-4a7b-479d-990f-5b639624c649

depends on https://github.com/Infomaniak/android-pdfview/pull/17

### CODE REVIEW

File: app/src/main/java/com/infomaniak/drive/ui/fileList/preview/BitmapPrintDocumentAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    private var pageHeight: Int = 0
+    private var pageWidth: Int = 0
+    private var pdfDocument: PrintedPdfDocument? = null
+
+    override fun onLayout(
+        oldAttributes: PrintAttributes?,
+        newAttributes: PrintAttributes?,
+        cancellationSignal: CancellationSignal?,
+        callback: LayoutResultCallback?,
+        extras: Bundle?
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/preview/BitmapPrintDocumentAdapter.kt:49]:
```suggestion
        extras: Bundle?,
```

File: app/src/main/java/com/infomaniak/drive/ui/fileList/preview/BitmapPrintDocumentAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        pageHeight = newAttributes?.mediaSize?.heightMils ?: 0
+
+        pdfDocument = PrintedPdfDocument(context, newAttributes!!)
+
+        val printInfo = PrintDocumentInfo.Builder(fileName)
+            .setContentType(PrintDocumentInfo.CONTENT_TYPE_DOCUMENT)
+            .setPageCount(bitmaps.size)
+            .build()
+
+        callback?.onLayoutFinished(printInfo, true)
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/preview/BitmapPrintDocumentAdapter.kt:66]: Can you name this `true` parameter ?

File: app/src/main/java/com/infomaniak/drive/ui/fileList/preview/BitmapPrintDocumentAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    override fun onWrite(
+        pages: Array<PageRange>?,
+        destination: ParcelFileDescriptor?,
+        cancellationSignal: CancellationSignal?,
+        callback: WriteResultCallback?
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/preview/BitmapPrintDocumentAdapter.kt:73]:
```suggestion
        callback: WriteResultCallback?,
```

File: app/src/main/java/com/infomaniak/drive/ui/fileList/preview/BitmapPrintDocumentAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            return
+        }
+
+        bitmaps.forEachIndexed { index, bitmap ->
+            pdfDocument?.startPage(index)?.also { page ->
+                val contentRect = RectF(page.info.contentRect)
+                val matrix = getMatrix(bitmap.width, bitmap.height, contentRect)
+                matrix.postTranslate(contentRect.left, contentRect.top)
+                page.canvas.clipRect(contentRect)
+                page.canvas.drawBitmap(bitmap, matrix, null)
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/preview/BitmapPrintDocumentAdapter.kt:86]: Can you name this `null` parameter ?

File: app/src/main/java/com/infomaniak/drive/ui/fileList/preview/BitmapPrintDocumentAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+        runCatching {
+            pdfDocument?.writeTo(FileOutputStream(destination?.fileDescriptor))
+        }.onFailure { exception ->
+            callback?.onWriteFailed(exception.toString())
+            return
+        }.getOrElse {
+            pdfDocument?.close()
+            pdfDocument = null
+        }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/preview/BitmapPrintDocumentAdapter.kt:99]: I think we need to use `try catch finally` here.

File: app/src/main/java/com/infomaniak/drive/ui/fileList/preview/BitmapPrintDocumentAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+        try {
+            pdfDocument?.writeTo(FileOutputStream(destination?.fileDescriptor))
+        } catch(exception: Exception) {
+            callback?.onWriteFailed(exception.toString())
+            return
+        } finally {
+            pdfDocument?.close()
+            pdfDocument = null
+        }
```
[the remark this answers is not in the document]
Comment (tevincent) [reply app/src/main/java/com/infomaniak/drive/ui/fileList/preview/BitmapPrintDocumentAdapter.kt:99]: Yes.

File: app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewPDFFragment.kt [hunk 7]
```diff
@@ -111,7 +115,7 @@ class PreviewPDFFragment : PreviewFragment() {
         bigOpenWithButton.apply {
             isGone = true
             setOnClickListener {
-                if (isPasswordProtected) showPasswordDialog() else openWithClicked()
+                if (previewPDFHandler.isPasswordProtected) showPasswordDialog() else openWithClicked()
             }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewPDFFragment.kt:119]: Can't add suggestion.

Replace with
```kt
            setOnClickListener { if (previewPDFHandler.isPasswordProtected) showPasswordDialog() else openWithClicked() }
```

File: app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewPDFFragment.kt [hunk 8]
```diff
@@ -159,17 +170,23 @@ class PreviewPDFFragment : PreviewFragment() {
...
+                            shouldHidePrintOption(isGone = false)
                             binding.downloadLayout.root.isGone = true
                             dismissPasswordDialog()
                             updatePageNumber(totalPage = pageCount)
+                            pdfViewPrintListener = this@PreviewPDFFragment
                         }
                         onPageChange { currentPage, pageCount ->
                             updatePageNumber(
                                 currentPage = currentPage,
                                 totalPage = pageCount
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/preview/PreviewPDFFragment.kt:182]:
```suggestion
                                totalPage = pageCount,
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

**Review by KevinBoulongne — APPROVED** [review]

**Pull request merged**